File: src/study-schedule.dart
import "dart:io";

void main() {
    Program().run();
}

class Program {
    bool exit = false;
    bool isDeleting = false;
    Command cmd = Command("");
    String message = "";

    List<ScheduleItem> list = [
        ScheduleItem(331, "Hello, World!", "How are you?"),
        ScheduleItem(201, "Second Title", "")
    ];

    void run() {
        while (!exit) {
            draw();
            update();
        }
    }

    void draw() {
        UI.ClearScreen();

        UI.Text("┏━━ Study Schedule ━━━\n");
        UI.EmptyLine();

        if (list.isNotEmpty) {
            UI.Text("  List:\n");
            UI.ListView(list.length, (int i) {
                var item = list.elementAt(i);

                if (isDeleting) {
                    UI.Text("    [${i + 1}] ${item.title}\n");
                } else {
                    UI.Text("    ${item.title}\n");
                }

                if (item.desc.isNotEmpty) {
                    UI.Text("      ${item.desc}\n");
                    UI.EmptyLine();
                } else {
                    UI.EmptyLine();
                }
            });
        }

        UI.EmptyLine();
        UI.EmptyLine();

        if (message.isNotEmpty) {
            UI.Text(message);
        } else {
            UI.EmptyLine();
        }

        UI.Text(">> ");
    }

    void update() {
        message = "";

        cmd = Command(stdin.readLineSync() ?? "");
        if (cmd.name.isEmpty) return;

        switch (cmd.name) {
            case "add":
            case "ad":
            case "a": {
                draw();
                stdout.write("Title (empty to cancel): ");
                var title = stdin.readLineSync() ?? "";
                title = title.trim();

                if (title.isEmpty) {
                    message = "Cancelled\n";
                    break;
                }

                stdout.write("Description (optional):\n");
                var desc = stdin.readLineSync() ?? "";
                desc = desc.trim();

                var item = ScheduleItem(
                    DateTime.now().microsecondsSinceEpoch,
                    title,
                    desc
                );

                list.add(item);
                message = "New item was successfully added\n";
            } break;

            case "delete":
            case "del":
            case "d": {
                if (list.isEmpty) {
                    message = "The list is empty. You can add items using `add` command\n";
                    break;
                }

                isDeleting = true;
                String input = "";
                int? index = null;

                while (index == null) {
                    draw();
                    stdout.write("Select item to delete (empty to cancel): ");

                    input = stdin.readLineSync() ?? "";
                    input = input.trim().toLowerCase();
                    if (input.isEmpty) break;

                    index = int.tryParse(input);
                    if (index == null) {
                        message = "Invalid input: `$input`\n";
                    } else if (index > list.length) {
                        message = "Item `[$index]` is not valid\n";
                        index = null;
                    }
                }

                if (input.isEmpty) {
                    isDeleting = false;
                    message = "Canceled\n";
                    break;
                }

                index = index! - 1;
                list.removeAt(index);

                isDeleting = false;
                message = "Item successfully deleted\n";
            } break;

            case "q":
            case "quit":
            case "exit":
                exit = true;
                UI.ClearScreen();
                break;

            default: message = "Unknown command: `${cmd.name}`\n";
        }
    }
}

class Command {
    String name = "";
    List<String> args = [];

    Command(String cmd) {
        cmd = cmd.trim().toLowerCase();

        if (cmd.isEmpty) return;

        args = cmd.split(" ")
            .where((x) => x.isNotEmpty)
            .toList();

        name = args.first;
        args.removeAt(0);
    }
}

class ScheduleItem {
    int id;
    String title;
    String desc;

    ScheduleItem(this.id, this.title, this.desc);
}

class UI {
    UI.ClearScreen() {
        stdout.write(Process.runSync("clear", []).stdout);
    }

    UI.Text(String text) {
        stdout.write(text);
    }

    UI.EmptyLine() {
        UI.Text("\n");
    }

    UI.ListView(int listLen, Function(int i) builder) {
        for (int i = 0; i < listLen; i++) {
            builder(i);
        }
    }
}

File: study-schedule.go
package main

import (
    "fmt"
    "os/exec"
    "os"
    "strings"
    "time"
    "bufio"
)

type Task struct {
    id int
    title string
    desc string
}

var exit bool = false
var isDeleting bool = false
var message string = ""
var tasks []Task = []Task{
    Task{id: 0, title: "Hello, World!", desc: "How are you?"},
    Task{id: 1, title: "Second Title", desc: ""},
}

var scanner *bufio.Scanner = bufio.NewScanner(os.Stdin)

func update() {
    message = ""

    scanner.Scan()
    cmd := strings.ToLower(strings.Trim(scanner.Text(), " "))
    if len(cmd) == 0 { return }

    switch cmd {
    case "add": {
        message = "Enter task title (empty for cancel):\n"
        draw()
        scanner.Scan()
        title := strings.Trim(scanner.Text(), " ")

        if len(title) > 0 {
            message = "Enter task description (optional):\n"
            draw()
            scanner.Scan()
            desc := strings.Trim(scanner.Text(), " ")

            id := time.Now().Nanosecond()
            newTask := Task{id, title, desc}
            tasks = append(tasks, newTask)
        }

        message = ""
    }

    case "delete":
        message = "TODO: Delete Item\n"

    case "q", "quit", "exit":
        exit = true
        clearScreen()

    default:
        message = fmt.Sprintf("Unknown command: `%s`\n", cmd)
    }
}

func draw() {
    clearScreen()

    drawText("┏━━ Study Schedule ━━━\n")
    drawEmptyLine()

    if len(tasks) > 0 {
        drawText("  List:\n")
        drawList(len(tasks), func (i int) {
            item := tasks[i]

            if isDeleting {
                drawText(fmt.Sprintf("    [%v] %v\n", i + 1, item.title))
            } else {
                drawText(fmt.Sprintf("    %v\n", item.title))
            }

            if (len(item.desc) > 0) {
                drawText("      " + item.desc + "\n")
            }
            drawEmptyLine()
        })
    }

    if len(message) > 0 {
        drawText(message)
    } else {
        drawEmptyLine()
    }

    drawText(">> ")
}

// UI
func clearScreen() {
    c := exec.Command("clear")
    c.Stdout = os.Stdout
    c.Run()
}

func drawText(s string) {
    fmt.Print(s)
}

func drawEmptyLine() {
    drawText("\n")
}

func drawList(listLen int, builder func(i int)) {
    for i := 0; i < listLen; i += 1 {
        builder(i)
    }
}

func main() {
    for !exit {
        draw()
        update()
    }
}
